### PR TITLE
Support rename collection

### PR DESCRIPTION
**Pull request by jaime0815**
issue  #21284
Signed-off-by: jaime <yun.zhang@zilliz.com>

### CODE REVIEW

File: internal/proxy/impl.go [hunk 1]
```diff
@@ -4407,6 +4407,35 @@ func (node *Proxy) CheckHealth(ctx context.Context, request *milvuspb.CheckHealt
 	}, nil
 }
 
+func (node *Proxy) RenameCollection(ctx context.Context, req *milvuspb.RenameCollectionRequest) (*commonpb.Status, error) {
+	ctx, sp := otel.Tracer(typeutil.ProxyRole).Start(ctx, "Proxy-RefreshPolicyInfoCache")
+	defer sp.End()
+
+	log := log.Ctx(ctx)
+	log.Debug("received rename collection request", zap.Any("request", req))
```
Comment (longjiquan) [inline internal/proxy/impl.go:4415]: - I think `Info` will be better here.
- Do not use `zap.Any` to print a struct, though maybe it's not large.

File: internal/proxy/impl.go [hunk 2]
```diff
@@ -4407,6 +4407,35 @@ func (node *Proxy) CheckHealth(ctx context.Context, request *milvuspb.CheckHealt
...
+	}
+
+	req.Base = commonpbutil.NewMsgBase(
+		commonpbutil.WithMsgType(commonpb.MsgType_RenameCollection),
+		commonpbutil.WithMsgID(0),
+		commonpbutil.WithSourceID(paramtable.GetNodeID()),
+	)
+	resp, err := node.rootCoord.RenameCollection(ctx, req)
+	if err != nil {
+		log.Warn("failed to rename collection", zap.Error(err))
```
Comment (longjiquan) [inline internal/proxy/impl.go:4429]: Better to print more detailed information about this request.

File: internal/rootcoord/meta_table.go [hunk 3]
```diff
@@ -22,32 +22,28 @@ import (
...
 )
 
-//go:generate mockery --name=IMetaTable --outpkg=mockrootcoord
+//go:generate mockery --name=IMetaTable --outpkg=mockrootcoord --filename=meta_table.go --with-expecter
 type IMetaTable interface {
 	AddCollection(ctx context.Context, coll *model.Collection) error
 	ChangeCollectionState(ctx context.Context, collectionID UniqueID, state pb.CollectionState, ts Timestamp) error
 	RemoveCollection(ctx context.Context, collectionID UniqueID, ts Timestamp) error
-	GetCollectionByName(ctx context.Context, collectionName string, ts Timestamp) (*model.Collection, error)
+	GetCollectionByName(ctx context.Context, collectionName string, ts Timestamp, allowUnavailable bool) (*model.Collection, error)
```
Comment (longjiquan) [inline internal/rootcoord/meta_table.go:46]: I think we should always filter unavailable for these requests by collection name.

File: internal/rootcoord/meta_table.go [hunk 4]
```diff
@@ -356,6 +353,11 @@ func (mt *MetaTable) GetCollectionByName(ctx context.Context, collectionName str
 	if err != nil {
 		return nil, err
 	}
+
+	if coll != nil && allowUnavailable {
```
Comment (longjiquan) [inline internal/rootcoord/meta_table.go:357]: Please cover this case if you necessarily need this feature. And do not use `Clone` here, it makes no sense.

File: internal/rootcoord/meta_table.go [hunk 5]
```diff
@@ -448,6 +450,30 @@ func (mt *MetaTable) AlterCollection(ctx context.Context, oldColl *model.Collect
 	return nil
 }
 
+func (mt *MetaTable) RenameCollection(ctx context.Context, oldName string, newName string, ts Timestamp) error {
+	mt.ddLock.RLock()
+	defer mt.ddLock.RUnlock()
+	ctx = contextutil.WithTenantID(ctx, Params.CommonCfg.ClusterName.GetValue())
+
+	oldColl, err := mt.GetCollectionByName(ctx, oldName, ts, true)
```
Comment (longjiquan) [inline internal/rootcoord/meta_table.go:458]: I don't think this behavior is reasonable. The users shouldn't see the unavailable collections.

Comment (longjiquan) [reply internal/rootcoord/meta_table.go:458]: Also, maybe this method will cause deadlock with `GetCollectionByName`.

Comment (longjiquan) [reply internal/rootcoord/meta_table.go:458]: By the way, we don't check if `oldName` is an alias of collection.

File: internal/proxy/impl.go [hunk 6]
```diff
@@ -4407,6 +4407,35 @@ func (node *Proxy) CheckHealth(ctx context.Context, request *milvuspb.CheckHealt
 	}, nil
 }
 
+func (node *Proxy) RenameCollection(ctx context.Context, req *milvuspb.RenameCollectionRequest) (*commonpb.Status, error) {
```
Comment (longjiquan) [inline internal/proxy/impl.go:4410]: One quick question: why we don't reuse the `AlterCollection` in the client?

File: internal/rootcoord/root_coord.go [hunk 7]
```diff
@@ -2495,6 +2495,58 @@ func (c *Core) ListPolicy(ctx context.Context, in *internalpb.ListPolicyRequest)
...
+		baseTask: baseTask{
+			ctx:  ctx,
+			core: c,
+			done: make(chan error, 1),
+		},
+		Req: req,
+	}
+
+	if err := c.scheduler.AddTask(t); err != nil {
+		log.Warn("failed to enqueue request to rename collection",
```
Comment (longjiquan) [inline internal/rootcoord/root_coord.go:2521]: Please print more detailed information about this request.

File: internal/rootcoord/root_coord.go [hunk 8]
```diff
@@ -2495,6 +2495,58 @@ func (c *Core) ListPolicy(ctx context.Context, in *internalpb.ListPolicyRequest)
...
+			zap.String("role", typeutil.RootCoordRole),
+			zap.Error(err))
+
+		metrics.RootCoordDDLReqCounter.WithLabelValues("RenameCollection", metrics.FailLabel).Inc()
+		return failStatus(commonpb.ErrorCode_UnexpectedError, err.Error()), nil
+	}
+
+	if err := t.WaitToFinish(); err != nil {
+		log.Warn("failed to rename collection",
```
Comment (longjiquan) [inline internal/rootcoord/root_coord.go:2530]: Please print more detailed information about this request.

File: internal/rootcoord/root_coord.go [hunk 9]
```diff
@@ -2495,6 +2495,58 @@ func (c *Core) ListPolicy(ctx context.Context, in *internalpb.ListPolicyRequest)
...
+			zap.Error(err))
+
+		metrics.RootCoordDDLReqCounter.WithLabelValues("RenameCollection", metrics.FailLabel).Inc()
+		return failStatus(commonpb.ErrorCode_UnexpectedError, err.Error()), nil
+	}
+
+	metrics.RootCoordDDLReqCounter.WithLabelValues("RenameCollection", metrics.SuccessLabel).Inc()
+	metrics.RootCoordDDLReqLatency.WithLabelValues("RenameCollection").Observe(float64(tr.ElapseSpan().Milliseconds()))
+
+	log.Info("rename collection done",
```
Comment (longjiquan) [inline internal/rootcoord/root_coord.go:2542]: Not very important, but can you unify the log to "done to rename collection"? I am used to doing filtering by this when I tried to investigate logs.

**Review by longjiquan — CHANGES_REQUESTED** [review]

File: internal/rootcoord/meta_table.go (hunk 5 above)
Comment (jaime0815) [reply internal/rootcoord/meta_table.go:458]: It just gets a collection meta by a collection name, it should not care about whether unavailable

File: internal/rootcoord/meta_table.go [hunk 10]
```diff
@@ -448,6 +450,30 @@ func (mt *MetaTable) AlterCollection(ctx context.Context, oldColl *model.Collect
...
+	if err != nil {
+		return err
+	}
+
+	newColl := oldColl.Clone()
+	newColl.Name = newName
```
Comment (longjiquan) [inline internal/rootcoord/meta_table.go:464]: We didn't check if `newName` is valid here:
- Whether `newName` satisfies the rule of collection?
- Whether `newName` exists or it's an alias of collection?

**Review by longjiquan — CHANGES_REQUESTED** [review]

File: internal/rootcoord/meta_table.go (hunk 3 above)
Comment (jaime0815) [reply internal/rootcoord/meta_table.go:46]: It just gets a collection meta by collection name,  and it is similar to the GetCollectionByID method

File: internal/rootcoord/meta_table.go [hunk 11]
```diff
@@ -448,6 +446,50 @@ func (mt *MetaTable) AlterCollection(ctx context.Context, oldColl *model.Collect
...
+	ctx = contextutil.WithTenantID(ctx, Params.CommonCfg.ClusterName.GetValue())
+
+	log := log.Ctx(ctx).With(zap.String("oldName", oldName), zap.String("newName", newName))
+
+	//old collection should not be an alias
+	_, ok := mt.collAlias2ID[oldName]
+	if ok {
+		log.Warn("unsupported use a alias to rename collection")
+		return fmt.Errorf("unsupported use an alias to rename collection, alias:%s", oldName)
+	}
```
Comment (congqixia) [inline internal/rootcoord/meta_table.go:461]: Shall we check `newName` is an alias here? What shall happen if user rename a collection to an existing alias?

Comment (jaime0815) [reply internal/rootcoord/meta_table.go:461]: Yes,  newName also is checked if it is an existing alias or collection name by GetCollectionByName

**Review by congqixia — APPROVED** [review]
/lgtm

**Pull request merged**